### PR TITLE
Added clone current vehicle feature.

### PR DESCRIPTION
**Pull request by aa15032261**
- Fixed some bugs in get_owned_mods_from_vehicle function.
- Fixed some weaponized PVs can't be cloned correctly.

image

### CODE REVIEW

**Review by Yimura — DISMISSED** [review]
Looks good to me, will have to test this later on.

**Review by Yimura — DISMISSED** [review]

**Comment by mentolixite** [comment]
This seems to cause a crash when spawning vehicles sometimes.

**Comment by aa15032261** [comment]
> This seems to cause a crash when spawning vehicles sometimes.

I’ll look into it

> This seems to cause a crash when spawning vehicles sometimes.

Clone pv or clone current?

**Comment by aa15032261** [comment]
> This seems to cause a crash when spawning vehicles sometimes.

Does the console show any error?

**Comment by aa15032261** [comment]
> This seems to cause a crash when spawning vehicles sometimes.

I've modified the code so it checks if the vehicle handle is valid before doing anything further. Hopefully it will fix the crash.

File: BigBaseV2/src/views/vehicle/view_pv.cpp [hunk 1]
```diff
@@ -111,41 +111,62 @@ namespace big
...
+
+								for (auto& [i, val] : veh_data)
 								{
-									y_offset = 5.f;
+									LOG(WARNING) << i << ": " << val;
 								}
 
-								auto spawn_location = ENTITY::GET_OFFSET_FROM_ENTITY_IN_WORLD_COORDS(self::ped, 0.f, y_offset, 0.f);
-								float spawn_heading = ENTITY::GET_ENTITY_HEADING(self::ped);
+								LOG(WARNING) << "A====";
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_pv.cpp]: Debug log

File: BigBaseV2/src/views/vehicle/view_pv.cpp [hunk 2]
```diff
@@ -111,41 +111,62 @@ namespace big
...
 
-								const char* spawn_plate = plate;
-								if (g->clone_pv.clone_plate)
+								for (auto& [i, val] : owned_mods)
 								{
-									spawn_plate = personal_veh->get_plate();
+									LOG(WARNING) << i << ": " << val;
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_pv.cpp]: Debug log

File: BigBaseV2/src/views/vehicle/view_pv.cpp [hunk 3]
```diff
@@ -111,41 +111,62 @@ namespace big
...
 								}
 
-								auto veh = vehicle::clone_from_vehicle_data(veh_data, spawn_location, spawn_heading);
+								LOG(WARNING) << "B====";
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_pv.cpp]: Debug log

File: BigBaseV2/src/views/vehicle/view_pv.cpp [hunk 4]
```diff
@@ -111,41 +111,62 @@ namespace big
...
 
-								if (g->clone_pv.spawn_maxed)
+								for (auto& [i, val] : owned_mods2)
 								{
-									vehicle::max_vehicle(veh);
+									LOG(WARNING) << i << ": " << val;
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_pv.cpp]: Debug log

File: BigBaseV2/src/views/vehicle/view_pv.cpp [hunk 5]
```diff
@@ -111,41 +111,62 @@ namespace big
...
 								}
 
-								vehicle::set_plate(veh, spawn_plate);
+								LOG(WARNING) << "C====";
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_pv.cpp]: Debug log

**Review by Yimura — CHANGES_REQUESTED** [review]
If you're still working on the PR you can change to a draft, that way I know that it isn't ready to review yet.

image

**Comment by aa15032261** [comment]
@Yimura This PR is finalized. Sorry for the trouble.

**Comment by mentolixite** [comment]
Appears to work fine, thanks!

**Comment by howdoiusekeyboard** [comment]
I have found some bugs
1. After u preview current vehicle in the spawn tab, the other vehicles also have the same preview irrespective of model.
2. In the PV tab, after u preview a pv and hover off, and then hover on the same preview, the preview appears with vehicle mods not there. Exact bug with current vehicle in spawn tab
3. In your PV 's, if you have multiple builds of a model, with different mods on, and preview one of the build, and then try to preview the other build, the other build is previewed with the same mods on even though it has different mods.
4. I don't know if this is intentional but if u enabled spawn maxed in spawn, the vehicles aren't previewed with the respective mods. Same thing with spawn maxed in PV tab (after u enable spawn clone)
5. Clone doensn't get previewed with the PV plate if u have clone PV plate enabled
6. If you spawn a pv while keeping the preview, the preview gets the mods set on removed after it spawns

**Comment by aa15032261** [comment]
> I have found some bugs
> 
> 1. After u preview current vehicle in the spawn tab, the other vehicles also have the same preview irrespective of model.
> 2. In the PV tab, after u preview a pv and hover off, and then hover on the same preview, the preview appears with vehicle mods not there. Exact bug with current vehicle in spawn tab
> 3. In your PV 's, if you have multiple builds of a model, with different mods on, and preview one of the build, and then try to preview the other build, the other build is previewed with the same mods on even though it has different mods.
> 4. I don't know if this is intentional but if u enabled spawn maxed in spawn, the vehicles aren't previewed with the respective mods. Same thing with spawn maxed in PV tab (after u enable spawn clone)
> 5. Clone doensn't get previewed with the PV plate if u have clone PV plate enabled
> 6. If you spawn a pv while keeping the preview, the preview gets the mods set on removed after it spawns

Thanks for the feedback! I have fixed all issues except 5 in the latest commit. I think the plate number is not that important so I left that out. Please let me know if you find other issues.

**Comment by howdoiusekeyboard** [comment]
Outstanding work man! Thanks

**Comment by aa15032261** [comment]
The latest commit added low grip option and fixed issue #362.

File: BigBaseV2/src/views/vehicle/view_spawn.cpp [hunk 6]
```diff
@@ -2,6 +2,7 @@
 #include "fiber_pool.hpp"
 #include "natives.hpp"
 #include "services/vehicle_preview/vehicle_preview_service.hpp"
+#include "services/players/player_service.hpp"
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_spawn.cpp]: What's this include used for?

**Review by Yimura — CHANGES_REQUESTED** [review]
Your vendor isn't up-to-date with master, you'll have to fix this as well before we can merge this.

File: BigBaseV2/src/views/vehicle/view_spawn.cpp (hunk 6 above)
Comment (aa15032261) [reply BigBaseV2/src/views/vehicle/view_spawn.cpp]: It was used in the previous version and I forgot to remove it. Thanks for catching that.

**Review by Yimura — APPROVED** [review]
Alright looks good to me, will test it real quick.

**Pull request merged**